app/controllers/AppealsController.scala

```diff
@@ -36,32 +37,43 @@ class AppealsController @Inject()(appConfig: AppConfig,
                                   cc: ControllerComponents)(implicit ec: ExecutionContext)
   extends BackendController(cc) {
 
+  private def getAppealDataForPenalty(penaltyId: String, enrolmentKey: String, penaltyType: AppealTypeEnum.Value)(implicit hc: HeaderCarrier): Future[Result] = {
+    etmpService.getPenaltyDataFromETMPForEnrolment(enrolmentKey).map {
+      result => {
+        result._1.fold {
+          result._2 match {
+            case Left(GetETMPPayloadNoContent) => NotFound(s"Could not retrieve ETMP penalty data for $enrolmentKey")
+            case Left(_) => InternalServerError("Something went wrong.")
+          }
+        }(
+          etmpData => {
+            checkAndReturnResponseForPenaltyData(etmpData, penaltyId, enrolmentKey, penaltyType)
+          }
+        )
+      }
+    }
+  }
+
   def getAppealsDataForLateSubmissionPenalty(penaltyId: String, enrolmentKey: String): Action[AnyContent] = Action.async {
     implicit request => {
-      etmpService.getPenaltyDataFromETMPForEnrolment(enrolmentKey).map {
-        result => {
-          result._1.fold {
-            result._2 match {
-              case Left(GetETMPPayloadNoContent) => NotFound(s"Could not retrieve ETMP penalty data for $enrolmentKey")
-              case Left(_) => InternalServerError("Something went wrong.")
-            }
-          }(
-            etmpData => {
-              checkAndReturnResponseForPenaltyData(etmpData, penaltyId, enrolmentKey, Late_Submission)
-            }
-          )
-        }
-      }
+      getAppealDataForPenalty(penaltyId, enrolmentKey, Late_Submission)
+    }
+  }
+
+  def getAppealsDataForLatePaymentPenalty(penaltyId: String, enrolmentKey: String): Action[AnyContent] = Action.async {
+    implicit request => {
+      getAppealDataForPenalty(penaltyId, enrolmentKey, Late_Payment)
     }
   }
 
   private def checkAndReturnResponseForPenaltyData(etmpData: ETMPPayload,
                                                    penaltyIdToCheck: String,
                                                    enrolmentKey: String,
                                                    appealType: AppealTypeEnum.Value): Result = {
-    val isPenaltyIdInETMPPayload: Boolean = etmpData.penaltyPoints.exists(_.id == penaltyIdToCheck)
-    if (isPenaltyIdInETMPPayload) {
-      logger.debug(s"[AppealsController][getAppealsData] Penalty ID: $penaltyIdToCheck for enrolment key: $enrolmentKey found in ETMP.")
```

Comment:
Looks like you are using the boolean check (_.id == penaltyIdToCheck) at five different places, could this be optimised.

Reply:
Makes sense, looking now 😄